src/core/MIGRATION_EXAMPLES.md

```diff
@@ -0,0 +1,255 @@
+import Joi from 'joi';
+
+new kibana.Plugin({
+  config() {
+    return Joi.object({
+      enabled: Joi.boolean().default(true),
+      defaultAppId: Joi.string().default('home'),
+      index: Joi.string().default('.kibana'),
+      disableWelcomeScreen: Joi.boolean().default(false),
+      autocompleteTerminateAfter: Joi.number().integer().min(1).default(100000),
+    })
+  }
+});
+
+// New Platform equivalent
+import { schema } from '@kbn/config-schema';
+
+// This object should be exported from your server-side plugin's index.ts file.
+export const config = {
+  schema: schema.object({
+    enabled: schema.boolean({ defaultValue: true }),
+    defaultAppId: schema.string({ defaultValue: true }),
+    index: schema.string({ defaultValue: '.kibana' }),
+    disableWelcomeScreen: schema.boolean({ defaultValue: false }),
+    autocompleteTerminateAfter: schema.number({ min: 1, defaultValue: 100000 }),
+  })
+};
+
+// @kbn/config-schema is written in TypeScript, so you can use your schema
+// definition to give you a type to use in your plugin code
+type MyPluginConfig = typeof config.schema
+```
+
+### Using New Platform config from a Legacy plugin
+
+During the migration process, you'll want to migrate your schema to the new
+format. However, legacy plugins cannot directly get access to New Platform's
+config service due to the way that config is tied to the `kibana.json` file
+(which does not exist for legacy plugins).
+
+There is a workaround though:
+- Create a New Platform plugin that contains your plugin's config schema in the new format
+- Expose the config from the New Platform plugin in its setup contract
+- Read the config from the setup contract in your legacy plugin
+
+#### Create a New Platform plugin
+
+For example, if wanted to move the legacy `timelion` plugin's configuration to
+the New Platform, we could create a NP plugin with the same name in
+`src/plugins/timelion` with the following files:
+
+```json5
+// src/plugins/timelion/kibana.json
+{
+  "id": "timelion",
+  "server": true
+}
+```
+
+```ts
+// src/plugins/timelion/server/index.ts
+import { schema } from '@kbn/config-schema';
+import { PluginInitializerContext } from '../../core/server';
+import { TimelionPlugin } from './plugin';
+
+export const config = {
+  schema: schema.object({
+    enabled: schema.boolean({ defaultValue: true }),
+  });
+}
+
+export const plugin = (initContext: PluginInitializerContext) => new TimelionPlugin(initContext);
+
+export type TimelionConfig = typeof config.schema;
+export { TimelionSetup } from './plugin';
+```
+
+```ts
+// src/plugins/timelion/server/plugin.ts
+import { PluginInitializerContext, Plugin, CoreSetup } from '../../core/server';
+import { TimelionConfig } from '.';
+
+export class TimelionPlugin implements Plugin<TimelionSetup> {
+  constructor(private readonly initContext: PluginInitializerContext) {}
+
+  public setup(core: CoreSetup) {
+    return {
+      config$: this.initContext.config.create<TimelionConfig>(),
```

Comment:
`config$` won't be a part of the public contract when plugin migrated. I'd suggest adding a convention for temporarily exposed API. We can place it under a special namespace. `legacy`, `__legacy`, etc. and mark as `@deprecated` no-one starts using it.